File: test/Models/pheno_pkg/src/cpp/PhenologyComponent.cpp
#include "PhenologyComponent.h"

    PhenologyComponent::PhenologyComponent()
    {
           
    }
    

float PhenologyComponent::getaMXLFNO() {return this-> aMXLFNO; }
float PhenologyComponent::getpNini() {return this-> pNini; }
float PhenologyComponent::getsDsa_sh() {return this-> sDsa_sh; }
float PhenologyComponent::getlatitude() {return this-> latitude; }
float PhenologyComponent::getkl() {return this-> kl; }
float PhenologyComponent::getlincr() {return this-> lincr; }
float PhenologyComponent::getldecr() {return this-> ldecr; }
float PhenologyComponent::getpincr() {return this-> pincr; }
float PhenologyComponent::getpTQhf() {return this-> pTQhf; }
float PhenologyComponent::getB() {return this-> B; }
float PhenologyComponent::getareaSL() {return this-> areaSL; }
float PhenologyComponent::getareaSS() {return this-> areaSS; }
float PhenologyComponent::getlARmin() {return this-> lARmin; }
float PhenologyComponent::getsowingDensity() {return this-> sowingDensity; }
float PhenologyComponent::getlARmax() {return this-> lARmax; }
float PhenologyComponent::getlNeff() {return this-> lNeff; }
float PhenologyComponent::getrp() {return this-> rp; }
float PhenologyComponent::getp() {return this-> p; }
float PhenologyComponent::getpdecr() {return this-> pdecr; }
float PhenologyComponent::getmaxTvern() {return this-> maxTvern; }
float PhenologyComponent::gettTWindowForPTQ() {return this-> tTWindowForPTQ; }
float PhenologyComponent::getvBEE() {return this-> vBEE; }
int PhenologyComponent::getisVernalizable() {return this-> isVernalizable; }
float PhenologyComponent::getminTvern() {return this-> minTvern; }
float PhenologyComponent::getintTvern() {return this-> intTvern; }
float PhenologyComponent::getvAI() {return this-> vAI; }
float PhenologyComponent::getmaxDL() {return this-> maxDL; }
string PhenologyComponent::getchoosePhyllUse() {return this-> choosePhyllUse; }
float PhenologyComponent::getminDL() {return this-> minDL; }
float PhenologyComponent::getpFLLAnth() {return this-> pFLLAnth; }
float PhenologyComponent::getdcd() {return this-> dcd; }
float PhenologyComponent::getdgf() {return this-> dgf; }
float PhenologyComponent::getdegfm() {return this-> degfm; }
bool PhenologyComponent::getignoreGrainMaturation() {return this-> ignoreGrainMaturation; }
float PhenologyComponent::getpHEADANTH() {return this-> pHEADANTH; }
float PhenologyComponent::getsLDL() {return this-> sLDL; }
int PhenologyComponent::getsowingDay() {return this-> sowingDay; }
string PhenologyComponent::getsowingDate() {return this-> sowingDate; }
int PhenologyComponent::getsDws() {return this-> sDws; }
float PhenologyComponent::getsDsa_nh() {return this-> sDsa_nh; }
float PhenologyComponent::getder() {return this-> der; }
float PhenologyComponent::gettargetFertileShoot() {return this-> targetFertileShoot; }
float PhenologyComponent::getdse() {return this-> dse; }
float PhenologyComponent::getslopeTSFLN() {return this-> slopeTSFLN; }
float PhenologyComponent::getintTSFLN() {return this-> intTSFLN; }

void PhenologyComponent::setaMXLFNO(float _aMXLFNO)
{
    _Vernalizationprogress.setaMXLFNO(_aMXLFNO);
}
void PhenologyComponent::setpNini(float _pNini)
{
    _Vernalizationprogress.setpNini(_pNini);
}
void PhenologyComponent::setsDsa_sh(float _sDsa_sh)
{
    _Phylsowingdatecorrection.setsDsa_sh(_sDsa_sh);
}
void PhenologyComponent::setlatitude(float _latitude)
{
    _Phylsowingdatecorrection.setlatitude(_latitude);
}
void PhenologyComponent::setkl(float _kl)
{
    _Phyllochron.setkl(_kl);
    _Ptq.setkl(_kl);
}
void PhenologyComponent::setlincr(float _lincr)
{
    _Phyllochron.setlincr(_lincr);
}
void PhenologyComponent::setldecr(float _ldecr)
{
    _Phyllochron.setldecr(_ldecr);
}
void PhenologyComponent::setpincr(float _pincr)
{
    _Phyllochron.setpincr(_pincr);
}
void PhenologyComponent::setpTQhf(float _pTQhf)
{
    _Phyllochron.setpTQhf(_pTQhf);
}
void PhenologyComponent::setB(float _B)
{
    _Phyllochron.setB(_B);
}
void PhenologyComponent::setareaSL(float _areaSL)
{
    _Phyllochron.setareaSL(_areaSL);
}
void PhenologyComponent::setareaSS(float _areaSS)
{
    _Phyllochron.setareaSS(_areaSS);
}
void PhenologyComponent::setlARmin(float _lARmin)
{
    _Phyllochron.setlARmin(_lARmin);
}
void PhenologyComponent::setsowingDensity(float _sowingDensity)
{
    _Phyllochron.setsowingDensity(_sowingDensity);
    _Shootnumber.setsowingDensity(_sowingDensity);
}
void PhenologyComponent::setlARmax(float _lARmax)
{
    _Phyllochron.setlARmax(_lARmax);
}
void PhenologyComponent::setlNeff(float _lNeff)
{
    _Phyllochron.setlNeff(_lNeff);
}
void PhenologyComponent::setrp(float _rp)
{
    _Phylsowingdatecorrection.setrp(_rp);
}
void PhenologyComponent::setp(float _p)
{
    _Phyllochron.setp(_p);
    _Phylsowingdatecorrection.setp(_p);
    _Updatephase.setp(_p);
}
void PhenologyComponent::setpdecr(float _pdecr)
{
    _Phyllochron.setpdecr(_pdecr);
}
void PhenologyComponent::setmaxTvern(float _maxTvern)
{
    _Vernalizationprogress.setmaxTvern(_maxTvern);
}
void PhenologyComponent::settTWindowForPTQ(float _tTWindowForPTQ)
{
    _Gaimean.settTWindowForPTQ(_tTWindowForPTQ);
    _Ptq.settTWindowForPTQ(_tTWindowForPTQ);
}
void PhenologyComponent::setvBEE(float _vBEE)
{
    _Vernalizationprogress.setvBEE(_vBEE);
}
void PhenologyComponent::setisVernalizable(int _isVernalizable)
{
    _Vernalizationprogress.setisVernalizable(_isVernalizable);
    _Updatephase.setisVernalizable(_isVernalizable);
}
void PhenologyComponent::setminTvern(float _minTvern)
{
    _Vernalizationprogress.setminTvern(_minTvern);
}
void PhenologyComponent::setintTvern(float _intTvern)
{
    _Vernalizationprogress.setintTvern(_intTvern);
}
void PhenologyComponent::setvAI(float _vAI)
{
    _Vernalizationprogress.setvAI(_vAI);
}
void PhenologyComponent::setmaxDL(float _maxDL)
{
    _Vernalizationprogress.setmaxDL(_maxDL);
    _Updatephase.setmaxDL(_maxDL);
}
void PhenologyComponent::setchoosePhyllUse(string _choosePhyllUse)
{
    _Phyllochron.setchoosePhyllUse(_choosePhyllUse);
    _Updatephase.setchoosePhyllUse(_choosePhyllUse);
}
void PhenologyComponent::setminDL(float _minDL)
{
    _Vernalizationprogress.setminDL(_minDL);
}
void PhenologyComponent::setpFLLAnth(float _pFLLAnth)
{
    _Updatephase.setpFLLAnth(_pFLLAnth);
}
void PhenologyComponent::setdcd(float _dcd)
{
    _Updatephase.setdcd(_dcd);
}
void PhenologyComponent::setdgf(float _dgf)
{
    _Updatephase.setdgf(_dgf);
}
void PhenologyComponent::setdegfm(float _degfm)
{
    _Updatephase.setdegfm(_degfm);
}
void PhenologyComponent::setignoreGrainMaturation(bool _ignoreGrainMaturation)
{
    _Updatephase.setignoreGrainMaturation(_ignoreGrainMaturation);
}
void PhenologyComponent::setpHEADANTH(float _pHEADANTH)
{
    _Updatephase.setpHEADANTH(_pHEADANTH);
}
void PhenologyComponent::setsLDL(float _sLDL)
{
    _Updatephase.setsLDL(_sLDL);
}
void PhenologyComponent::setsowingDay(int _sowingDay)
{
    _Phylsowingdatecorrection.setsowingDay(_sowingDay);
}
void PhenologyComponent::setsowingDate(string _sowingDate)
{
    _Registerzadok.setsowingDate(_sowingDate);
}
void PhenologyComponent::setsDws(int _sDws)
{
    _Phylsowingdatecorrection.setsDws(_sDws);
}
void PhenologyComponent::setsDsa_nh(float _sDsa_nh)
{
    _Phylsowingdatecorrection.setsDsa_nh(_sDsa_nh);
}
void PhenologyComponent::setder(float _der)
{
    _Registerzadok.setder(_der);
}
void PhenologyComponent::settargetFertileShoot(float _targetFertileShoot)
{
    _Shootnumber.settargetFertileShoot(_targetFertileShoot);
}
void PhenologyComponent::setdse(float _dse)
{
    _Updatephase.setdse(_dse);
}
void PhenologyComponent::setslopeTSFLN(float _slopeTSFLN)
{
    _Registerzadok.setslopeTSFLN(_slopeTSFLN);
}
void PhenologyComponent::setintTSFLN(float _intTSFLN)
{
    _Registerzadok.setintTSFLN(_intTSFLN);
}
void PhenologyComponent::Calculate_Model(PhenologyState& s, PhenologyState& s1, PhenologyRate& r, PhenologyAuxiliary& a)
{
    _Gaimean.Calculate_Model(s, s1, r, a);
    _Ptq.Calculate_Model(s, s1, r, a);
    _Cumulttfrom.Calculate_Model(s, s1, r, a);
    _Ismomentregistredzc_39.Calculate_Model(s, s1, r, a);
    _Vernalizationprogress.Calculate_Model(s, s1, r, a);
    _Phylsowingdatecorrection.Calculate_Model(s, s1, r, a);
    _Updatephase.Calculate_Model(s, s1, r, a);
    _Leafnumber.Calculate_Model(s, s1, r, a);
    _Shootnumber.Calculate_Model(s, s1, r, a);
    _Updateleafflag.Calculate_Model(s, s1, r, a);
    _Registerzadok.Calculate_Model(s, s1, r, a);
    _Updatecalendar.Calculate_Model(s, s1, r, a);
    _Phyllochron.Calculate_Model(s, s1, r, a);
}
PhenologyComponent::PhenologyComponent(const PhenologyComponent& toCopy)
{
    aMXLFNO = toCopy.aMXLFNO;
    pNini = toCopy.pNini;
    sDsa_sh = toCopy.sDsa_sh;
    latitude = toCopy.latitude;
    kl = toCopy.kl;
    lincr = toCopy.lincr;
    ldecr = toCopy.ldecr;
    pincr = toCopy.pincr;
    pTQhf = toCopy.pTQhf;
    B = toCopy.B;
    areaSL = toCopy.areaSL;
    areaSS = toCopy.areaSS;
    lARmin = toCopy.lARmin;
    sowingDensity = toCopy.sowingDensity;
    lARmax = toCopy.lARmax;
    lNeff = toCopy.lNeff;
    rp = toCopy.rp;
    p = toCopy.p;
    pdecr = toCopy.pdecr;
    maxTvern = toCopy.maxTvern;
    tTWindowForPTQ = toCopy.tTWindowForPTQ;
    vBEE = toCopy.vBEE;
    isVernalizable = toCopy.isVernalizable;
    minTvern = toCopy.minTvern;
    intTvern = toCopy.intTvern;
    vAI = toCopy.vAI;
    maxDL = toCopy.maxDL;
    choosePhyllUse = toCopy.choosePhyllUse;
    minDL = toCopy.minDL;
    pFLLAnth = toCopy.pFLLAnth;
    dcd = toCopy.dcd;
    dgf = toCopy.dgf;
    degfm = toCopy.degfm;
    ignoreGrainMaturation = toCopy.ignoreGrainMaturation;
    pHEADANTH = toCopy.pHEADANTH;
    sLDL = toCopy.sLDL;
    sowingDay = toCopy.sowingDay;
    sowingDate = toCopy.sowingDate;
    sDws = toCopy.sDws;
    sDsa_nh = toCopy.sDsa_nh;
    der = toCopy.der;
    targetFertileShoot = toCopy.targetFertileShoot;
    dse = toCopy.dse;
    slopeTSFLN = toCopy.slopeTSFLN;
    intTSFLN = toCopy.intTSFLN;
}


void PhenologyComponent::Init(PhenologyState& s, PhenologyState& s1, PhenologyRate& r, PhenologyAuxiliary& a)
{
    s.calendarMoments.push_back("Sowing");
    s.calendarCumuls.push_back(0.0f);
    s.calendarDates.push_back(sowingDate);
    s.minFinalNumber = 5.5f;
}

File: test/Models/pheno_pkg/src/java/PhenologyComponent.java
public class PhenologyComponent
{
    
    public PhenologyComponent() { }

    Phyllochron _Phyllochron = new Phyllochron();
    Phylsowingdatecorrection _Phylsowingdatecorrection = new Phylsowingdatecorrection();
    Shootnumber _Shootnumber = new Shootnumber();
    Updateleafflag _Updateleafflag = new Updateleafflag();
    Updatephase _Updatephase = new Updatephase();
    Leafnumber _Leafnumber = new Leafnumber();
    Vernalizationprogress _Vernalizationprogress = new Vernalizationprogress();
    Ismomentregistredzc_39 _Ismomentregistredzc_39 = new Ismomentregistredzc_39();
    Cumulttfrom _Cumulttfrom = new Cumulttfrom();
    Updatecalendar _Updatecalendar = new Updatecalendar();
    Registerzadok _Registerzadok = new Registerzadok();
    Ptq _Ptq = new Ptq();
    Gaimean _Gaimean = new Gaimean();

    public double getaMXLFNO()
    { return _Vernalizationprogress.getaMXLFNO(); }
    public void setaMXLFNO(double aMXLFNO)
    { _Vernalizationprogress.setaMXLFNO(aMXLFNO); } 

    public double getpNini()
    { return _Vernalizationprogress.getpNini(); }
    public void setpNini(double pNini)
    { _Vernalizationprogress.setpNini(pNini); } 

    public double getsDsa_sh()
    { return _Phylsowingdatecorrection.getsDsa_sh(); }
    public void setsDsa_sh(double sDsa_sh)
    { _Phylsowingdatecorrection.setsDsa_sh(sDsa_sh); } 

    public double getlatitude()
    { return _Phylsowingdatecorrection.getlatitude(); }
    public void setlatitude(double latitude)
    { _Phylsowingdatecorrection.setlatitude(latitude); } 

    public double getkl()
    { return _Phyllochron.getkl(); }
    public void setkl(double kl)
    { _Phyllochron.setkl(kl);
        _Ptq.setkl(kl); } 

    public double getlincr()
    { return _Phyllochron.getlincr(); }
    public void setlincr(double lincr)
    { _Phyllochron.setlincr(lincr); } 

    public double getldecr()
    { return _Phyllochron.getldecr(); }
    public void setldecr(double ldecr)
    { _Phyllochron.setldecr(ldecr); } 

    public double getpincr()
    { return _Phyllochron.getpincr(); }
    public void setpincr(double pincr)
    { _Phyllochron.setpincr(pincr); } 

    public double getpTQhf()
    { return _Phyllochron.getpTQhf(); }
    public void setpTQhf(double pTQhf)
    { _Phyllochron.setpTQhf(pTQhf); } 

    public double getB()
    { return _Phyllochron.getB(); }
    public void setB(double B)
    { _Phyllochron.setB(B); } 

    public double getareaSL()
    { return _Phyllochron.getareaSL(); }
    public void setareaSL(double areaSL)
    { _Phyllochron.setareaSL(areaSL); } 

    public double getareaSS()
    { return _Phyllochron.getareaSS(); }
    public void setareaSS(double areaSS)
    { _Phyllochron.setareaSS(areaSS); } 

    public double getlARmin()
    { return _Phyllochron.getlARmin(); }
    public void setlARmin(double lARmin)
    { _Phyllochron.setlARmin(lARmin); } 

    public double getsowingDensity()
    { return _Phyllochron.getsowingDensity(); }
    public void setsowingDensity(double sowingDensity)
    { _Phyllochron.setsowingDensity(sowingDensity);
        _Shootnumber.setsowingDensity(sowingDensity); } 

    public double getlARmax()
    { return _Phyllochron.getlARmax(); }
    public void setlARmax(double lARmax)
    { _Phyllochron.setlARmax(lARmax); } 

    public double getlNeff()
    { return _Phyllochron.getlNeff(); }
    public void setlNeff(double lNeff)
    { _Phyllochron.setlNeff(lNeff); } 

    public double getrp()
    { return _Phylsowingdatecorrection.getrp(); }
    public void setrp(double rp)
    { _Phylsowingdatecorrection.setrp(rp); } 

    public double getp()
    { return _Phyllochron.getp(); }
    public void setp(double p)
    { _Phyllochron.setp(p);
        _Phylsowingdatecorrection.setp(p);
        _Updatephase.setp(p); } 

    public double getpdecr()
    { return _Phyllochron.getpdecr(); }
    public void setpdecr(double pdecr)
    { _Phyllochron.setpdecr(pdecr); } 

    public double getmaxTvern()
    { return _Vernalizationprogress.getmaxTvern(); }
    public void setmaxTvern(double maxTvern)
    { _Vernalizationprogress.setmaxTvern(maxTvern); } 

    public double gettTWindowForPTQ()
    { return _Gaimean.gettTWindowForPTQ(); }
    public void settTWindowForPTQ(double tTWindowForPTQ)
    { _Gaimean.settTWindowForPTQ(tTWindowForPTQ);
        _Ptq.settTWindowForPTQ(tTWindowForPTQ); } 

    public double getvBEE()
    { return _Vernalizationprogress.getvBEE(); }
    public void setvBEE(double vBEE)
    { _Vernalizationprogress.setvBEE(vBEE); } 

    public int getisVernalizable()
    { return _Vernalizationprogress.getisVernalizable(); }
    public void setisVernalizable(int isVernalizable)
    { _Vernalizationprogress.setisVernalizable(isVernalizable);
        _Updatephase.setisVernalizable(isVernalizable); } 

    public double getminTvern()
    { return _Vernalizationprogress.getminTvern(); }
    public void setminTvern(double minTvern)
    { _Vernalizationprogress.setminTvern(minTvern); } 

    public double getintTvern()
    { return _Vernalizationprogress.getintTvern(); }
    public void setintTvern(double intTvern)
    { _Vernalizationprogress.setintTvern(intTvern); } 

    public double getvAI()
    { return _Vernalizationprogress.getvAI(); }
    public void setvAI(double vAI)
    { _Vernalizationprogress.setvAI(vAI); } 

    public double getmaxDL()
    { return _Vernalizationprogress.getmaxDL(); }
    public void setmaxDL(double maxDL)
    { _Vernalizationprogress.setmaxDL(maxDL);
        _Updatephase.setmaxDL(maxDL); } 

    public String getchoosePhyllUse()
    { return _Phyllochron.getchoosePhyllUse(); }
    public void setchoosePhyllUse(String choosePhyllUse)
    { _Phyllochron.setchoosePhyllUse(choosePhyllUse);
        _Updatephase.setchoosePhyllUse(choosePhyllUse); } 

    public double getminDL()
    { return _Vernalizationprogress.getminDL(); }
    public void setminDL(double minDL)
    { _Vernalizationprogress.setminDL(minDL); } 

    public double getpFLLAnth()
    { return _Updatephase.getpFLLAnth(); }
    public void setpFLLAnth(double pFLLAnth)
    { _Updatephase.setpFLLAnth(pFLLAnth); } 

    public double getdcd()
    { return _Updatephase.getdcd(); }
    public void setdcd(double dcd)
    { _Updatephase.setdcd(dcd); } 

    public double getdgf()
    { return _Updatephase.getdgf(); }
    public void setdgf(double dgf)
    { _Updatephase.setdgf(dgf); } 

    public double getdegfm()
    { return _Updatephase.getdegfm(); }
    public void setdegfm(double degfm)
    { _Updatephase.setdegfm(degfm); } 

    public boolean getignoreGrainMaturation()
    { return _Updatephase.getignoreGrainMaturation(); }
    public void setignoreGrainMaturation(boolean ignoreGrainMaturation)
    { _Updatephase.setignoreGrainMaturation(ignoreGrainMaturation); } 

    public double getpHEADANTH()
    { return _Updatephase.getpHEADANTH(); }
    public void setpHEADANTH(double pHEADANTH)
    { _Updatephase.setpHEADANTH(pHEADANTH); } 

    public double getsLDL()
    { return _Updatephase.getsLDL(); }
    public void setsLDL(double sLDL)
    { _Updatephase.setsLDL(sLDL); } 

    public int getsowingDay()
    { return _Phylsowingdatecorrection.getsowingDay(); }
    public void setsowingDay(int sowingDay)
    { _Phylsowingdatecorrection.setsowingDay(sowingDay); } 

    public LocalDateTime getsowingDate()
    { return _Registerzadok.getsowingDate(); }
    public void setsowingDate(LocalDateTime sowingDate)
    { _Registerzadok.setsowingDate(sowingDate); } 

    public int getsDws()
    { return _Phylsowingdatecorrection.getsDws(); }
    public void setsDws(int sDws)
    { _Phylsowingdatecorrection.setsDws(sDws); } 

    public double getsDsa_nh()
    { return _Phylsowingdatecorrection.getsDsa_nh(); }
    public void setsDsa_nh(double sDsa_nh)
    { _Phylsowingdatecorrection.setsDsa_nh(sDsa_nh); } 

    public double getder()
    { return _Registerzadok.getder(); }
    public void setder(double der)
    { _Registerzadok.setder(der); } 

    public double gettargetFertileShoot()
    { return _Shootnumber.gettargetFertileShoot(); }
    public void settargetFertileShoot(double targetFertileShoot)
    { _Shootnumber.settargetFertileShoot(targetFertileShoot); } 

    public double getdse()
    { return _Updatephase.getdse(); }
    public void setdse(double dse)
    { _Updatephase.setdse(dse); } 

    public double getslopeTSFLN()
    { return _Registerzadok.getslopeTSFLN(); }
    public void setslopeTSFLN(double slopeTSFLN)
    { _Registerzadok.setslopeTSFLN(slopeTSFLN); } 

    public double getintTSFLN()
    { return _Registerzadok.getintTSFLN(); }
    public void setintTSFLN(double intTSFLN)
    { _Registerzadok.setintTSFLN(intTSFLN); } 
    public void  Calculate_phenology(PhenologyState s, PhenologyState s1, PhenologyRate r, PhenologyAuxiliary a)
    {
        _Gaimean.Calculate_gaimean(s, s1, r, a);
        _Ptq.Calculate_ptq(s, s1, r, a);
        _Cumulttfrom.Calculate_cumulttfrom(s, s1, r, a);
        _Ismomentregistredzc_39.Calculate_ismomentregistredzc_39(s, s1, r, a);
        _Vernalizationprogress.Calculate_vernalizationprogress(s, s1, r, a);
        _Phylsowingdatecorrection.Calculate_phylsowingdatecorrection(s, s1, r, a);
        _Updatephase.Calculate_updatephase(s, s1, r, a);
        _Leafnumber.Calculate_leafnumber(s, s1, r, a);
        _Shootnumber.Calculate_shootnumber(s, s1, r, a);
        _Updateleafflag.Calculate_updateleafflag(s, s1, r, a);
        _Registerzadok.Calculate_registerzadok(s, s1, r, a);
        _Updatecalendar.Calculate_updatecalendar(s, s1, r, a);
        _Phyllochron.Calculate_phyllochron(s, s1, r, a);
    }
    private double aMXLFNO;
    private double pNini;
    private double sDsa_sh;
    private double latitude;
    private double kl;
    private double lincr;
    private double ldecr;
    private double pincr;
    private double pTQhf;
    private double B;
    private double areaSL;
    private double areaSS;
    private double lARmin;
    private double sowingDensity;
    private double lARmax;
    private double lNeff;
    private double rp;
    private double p;
    private double pdecr;
    private double maxTvern;
    private double tTWindowForPTQ;
    private double vBEE;
    private int isVernalizable;
    private double minTvern;
    private double intTvern;
    private double vAI;
    private double maxDL;
    private String choosePhyllUse;
    private double minDL;
    private double pFLLAnth;
    private double dcd;
    private double dgf;
    private double degfm;
    private boolean ignoreGrainMaturation;
    private double pHEADANTH;
    private double sLDL;
    private int sowingDay;
    private LocalDateTime sowingDate;
    private int sDws;
    private double sDsa_nh;
    private double der;
    private double targetFertileShoot;
    private double dse;
    private double slopeTSFLN;
    private double intTSFLN;
    public PhenologyComponent(PhenologyComponent toCopy) // copy constructor 
    {
        this.aMXLFNO = toCopy.getaMXLFNO();
        this.pNini = toCopy.getpNini();
        this.sDsa_sh = toCopy.getsDsa_sh();
        this.latitude = toCopy.getlatitude();
        this.kl = toCopy.getkl();
        this.lincr = toCopy.getlincr();
        this.ldecr = toCopy.getldecr();
        this.pincr = toCopy.getpincr();
        this.pTQhf = toCopy.getpTQhf();
        this.B = toCopy.getB();
        this.areaSL = toCopy.getareaSL();
        this.areaSS = toCopy.getareaSS();
        this.lARmin = toCopy.getlARmin();
        this.sowingDensity = toCopy.getsowingDensity();
        this.lARmax = toCopy.getlARmax();
        this.lNeff = toCopy.getlNeff();
        this.rp = toCopy.getrp();
        this.p = toCopy.getp();
        this.pdecr = toCopy.getpdecr();
        this.maxTvern = toCopy.getmaxTvern();
        this.tTWindowForPTQ = toCopy.gettTWindowForPTQ();
        this.vBEE = toCopy.getvBEE();
        this.isVernalizable = toCopy.getisVernalizable();
        this.minTvern = toCopy.getminTvern();
        this.intTvern = toCopy.getintTvern();
        this.vAI = toCopy.getvAI();
        this.maxDL = toCopy.getmaxDL();
        this.choosePhyllUse = toCopy.getchoosePhyllUse();
        this.minDL = toCopy.getminDL();
        this.pFLLAnth = toCopy.getpFLLAnth();
        this.dcd = toCopy.getdcd();
        this.dgf = toCopy.getdgf();
        this.degfm = toCopy.getdegfm();
        this.ignoreGrainMaturation = toCopy.getignoreGrainMaturation();
        this.pHEADANTH = toCopy.getpHEADANTH();
        this.sLDL = toCopy.getsLDL();
        this.sowingDay = toCopy.getsowingDay();
        this.sowingDate = toCopy.getsowingDate();
        this.sDws = toCopy.getsDws();
        this.sDsa_nh = toCopy.getsDsa_nh();
        this.der = toCopy.getder();
        this.targetFertileShoot = toCopy.gettargetFertileShoot();
        this.dse = toCopy.getdse();
        this.slopeTSFLN = toCopy.getslopeTSFLN();
        this.intTSFLN = toCopy.getintTSFLN();

    }
    

    

    public void Init(PhenologyState s, PhenologyState s1, PhenologyRate r, PhenologyAuxiliary a)
    {
        calendarMoments.add("Sowing");
        calendarCumuls.add(0.0d);
        calendarDates.add(sowingDate);
        minFinalNumber = 5.5d;
    }
    
}